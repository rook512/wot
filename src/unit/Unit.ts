import { DamageType } from "./damageType/DamageType";
import { UnitProperty } from "./unitProperty/UnitProperty";
import { Race } from "../races/Race";
import { Building } from "../buildings/Building";
import { UnitResource } from "./UnitResource";
import { BuildingResource } from "../buildings/BuildingResource";

export class Unit {
  readonly name: string;
  readonly requires: Array<Building | Race>;
  readonly cost: Record<BuildingResource, number>;
  readonly upkeep: Record<UnitResource, number>;
  readonly movement: number;
  readonly count: number;
  readonly maxHealth: number;
  readonly toHit: number;
  readonly damage: number;
  readonly damageType: Array<DamageType>;
  readonly rangedToHit: number;
  readonly rangedDamage: number;
  readonly rangedDamageType: Array<DamageType>;
  readonly ammo: number;
  readonly defense: number;
  readonly armor: number;
  readonly resistance: number;
  readonly unitProperty: Array<UnitProperty>;
  constructor(
    name: string,
    requires: Array<Building | Race>,
    cost: Record<BuildingResource, number>,
    upkeep: Record<UnitResource, number>,
    movement: number,
    count: number,
    maxHealth: number,
    toHit: number,
    damage: number,
    damageType: Array<DamageType>,
    rangedToHit: number,
    rangedDamage: number,
    rangedDamageType: Array<DamageType>,
    ammo: number,
    defense: number,
    armor: number,
    resistance: number,
    unitProperty: Array<UnitProperty>
  ) {
    this.name = name;
    this.requires = requires;
    this.cost = cost;
    this.upkeep = upkeep;
    this.movement = movement;
    this.count = count;
    this.maxHealth = maxHealth;
    this.toHit = toHit;
    this.damage = damage;
    this.damageType = damageType;
    this.rangedToHit = rangedToHit;
    this.rangedDamage = rangedDamage;
    this.rangedDamageType = rangedDamageType;
    this.ammo = ammo;
    this.defense = defense;
    this.armor = armor;
    this.resistance = resistance;
    this.unitProperty = unitProperty;
  }
}
