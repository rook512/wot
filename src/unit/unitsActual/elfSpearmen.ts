import { elf } from "../../races/elf/elf";
import { SpearmenBuilder } from "../unitTemplate/SpearmenBuilder";

export const elfSpearmen = new SpearmenBuilder()
  .setName("Elven Spearmen")
  .setRequires([elf])
  .setCount(4)
  .setToHit(2)
  .setDefense(2)
  .build();
