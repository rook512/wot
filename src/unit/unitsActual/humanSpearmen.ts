import { human } from "../../races/human/human";
import { SpearmenBuilder } from "../unitTemplate/SpearmenBuilder";

export const humanSpearmen = new SpearmenBuilder()
  .setName("Human Spearmen")
  .setRequires([human])
  .setCosts({ Production: 20 })
  .build();
