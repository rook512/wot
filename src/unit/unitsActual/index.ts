export * from "./aerokSpearmen";
export * from "./elfSpearmen";
export * from "./humanSpearmen";
export * from "./orcSpearmen";
