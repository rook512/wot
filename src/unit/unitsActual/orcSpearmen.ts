import { orc } from "../../races/orc/orc";
import { SpearmenBuilder } from "../unitTemplate/SpearmenBuilder";

export const orcSpearmen = new SpearmenBuilder()
  .setName("Orcish Spearmen")
  .setRequires([orc])
  .setCount(8)
  .build();
