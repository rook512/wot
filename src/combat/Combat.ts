import { Unit } from "../unit/Unit";

// export function combat(attacker: Unit, defender: Unit) {
//   let attacks = 0;
//   const attackAttempts = attacker.count * attacker.damage;
//   for (let attackNumber = 0; attackNumber < attackAttempts; attackNumber++) {
//     attacks = attacks + flipCoin();
//   }
//   let blocks = 0;
//   const blockAttempts = defender.count * defender.defense;
//   for (let blockNumber = 0; blockNumber < blockAttempts; blockNumber++) {
//     blocks = blocks + flipCoin();
//   }
//   const results: number = attacks - blocks;
//   return Math.max(0, results);
// }

function flipCoin(chanceOf1: number) {
  if (Math.random() > chanceOf1) {
    return 0;
  } else {
    return 1;
  }
}

export function combat(attacker: Unit, defender: Unit) {
  let defenseDifferential = calculateDefenseDifferential(attacker, defender);
  let attacks = 0;
  for (let attackNumber = 0; attackNumber < attacker.count; attackNumber++) {
    attacks = attacks + flipCoin(defenseDifferential);
  }
  const damage = attacks * attacker.damage;
  let defenderArmor = defender.count * defender.armor;
  let saves = 0;
  for (let savesNumber = 0; savesNumber < defenderArmor; savesNumber++) {
    saves = saves + flipCoin(0.3);
  }
  return Math.max(0, damage - saves);
}

export function calculateDefenseDifferential(attacker: Unit, defender: Unit) {
  const attackerTotal = attacker.count * attacker.toHit;
  const defenderTotal = defender.count * defender.defense;
  const advantageRatio = attackerTotal >= defenderTotal ? 0.1 : 0.1;
  let defenseDifferential =
    0.3 + (attackerTotal - defenderTotal) * advantageRatio;
  defenseDifferential = clamp(defenseDifferential, 0.1, 0.9);
  return defenseDifferential;
}

function clamp(original: number, min: number, max: number) {
  return Math.max(min, Math.min(max, original));
}
