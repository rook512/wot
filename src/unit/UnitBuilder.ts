import { Building } from "../buildings/Building";
import { Race } from "../races/Race";
import { UnitResource } from "./UnitResource";
import { BuildingResource } from "../buildings/BuildingResource";
import { DamageType } from "./damageType/DamageType";
import { UnitProperty } from "./unitProperty/UnitProperty";
import { Unit } from "./Unit";

export class UnitBuilder {
  private wasNameSet = false;
  name: string = "Unnamed Unit";
  requires: Array<Building | Race> = [];
  cost: Record<BuildingResource, number> = { Production: 0, Magic: 0, Holy: 0 };
  upkeep: Record<UnitResource, number> = {
    Money: 0,
    Food: 0,
    Magic: 0,
    Holy: 0
  };
  movement: number = 0;
  count: number = 0;
  maxHealth: number = 0;
  toHit: number = 0;
  damage: number = 0;
  damageType: Array<DamageType> = [];
  rangedToHit: number = 0;
  rangedDamage: number = 0;
  rangedDamageType: Array<DamageType> = [];
  ammo: number = 0;
  defense: number = 0;
  armor: number = 0;
  resistance: number = 0;
  unitProperty: Array<UnitProperty> = [];

  constructor(isTest = false) {
    this.wasNameSet = isTest;
  }

  setName(name: string) {
    this.wasNameSet = true;
    this.name = name;
    return this;
  }
  setRequires(requires: Array<Building | Race>) {
    this.requires = requires;
    return this;
  }
  setCosts(newCost: Partial<Record<BuildingResource, number>>) {
    this.cost = { ...this.cost, ...newCost };
    return this;
  }
  setUpkeep(newUpkeep: Partial<Record<UnitResource, number>>) {
    this.upkeep = { ...this.upkeep, ...newUpkeep };
    return this;
  }
  setMovement(movement: number) {
    this.movement = movement;
    return this;
  }
  setCount(count: number) {
    this.count = count;
    return this;
  }
  setMaxHealth(maxHealth: number) {
    this.maxHealth = maxHealth;
    return this;
  }
  setToHit(toHit: number) {
    this.toHit = toHit;
    return this;
  }
  setDamage(damage: number) {
    this.damage = damage;
    return this;
  }
  setDamageType(damageType: Array<DamageType>) {
    this.damageType = damageType;
    return this;
  }
  setRangedToHit(rangedToHit: number) {
    this.rangedToHit = rangedToHit;
    return this;
  }
  setRangedDamage(rangedDamage: number) {
    this.rangedDamage = rangedDamage;
    return this;
  }
  setRangedDamageType(rangedDamageType: Array<DamageType>) {
    this.rangedDamageType = rangedDamageType;
    return this;
  }
  setAmmo(ammo: number) {
    this.ammo = ammo;
    return this;
  }
  setDefense(defense: number) {
    this.defense = defense;
    return this;
  }
  setArmor(armor: number) {
    this.armor = armor;
    return this;
  }
  setResistance(resistance: number) {
    this.resistance = resistance;
    return this;
  }
  setUnitProperty(unitProperty: Array<UnitProperty>) {
    this.unitProperty = unitProperty;
    return this;
  }

  build(): Unit {
    if (!this.wasNameSet) {
      throw new Error("Could not build Unit; name was not set!");
    }
    return new Unit(
      this.name,
      this.requires,
      this.cost,
      this.upkeep,
      this.movement,
      this.count,
      this.maxHealth,
      this.toHit,
      this.damage,
      this.damageType,
      this.rangedToHit,
      this.rangedDamage,
      this.rangedDamageType,
      this.ammo,
      this.defense,
      this.armor,
      this.resistance,
      this.unitProperty
    );
  }
}
