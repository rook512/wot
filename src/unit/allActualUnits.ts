import { Unit } from "./Unit";

export const allActualUnits = (function(context) {
  try {
    const requireUnit = context();
    return (requireUnit.keys() as string[]).reduce(
      (prev: Unit[], key: string) => {
        prev.push(
          ...(Object.values(requireUnit(key)).filter(
            entry => entry instanceof Unit
          ) as Unit[])
        );
        return prev;
      },
      [] as Unit[]
    );
  } catch (ex) {
    return [];
  }
})(() => (require as any).context("./unitsActual", false, /\.ts$/));
