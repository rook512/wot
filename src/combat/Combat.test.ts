import { combat, calculateDefenseDifferential } from "./Combat";
import { humanSpearmen } from "../unit/unitsActual/humanSpearmen";
import { elfSpearmen } from "../unit/unitsActual/elfSpearmen";
import { orcSpearmen } from "../unit/unitsActual/orcSpearmen";
import { Unit } from "../unit/Unit";

function testCombatDifferential(attacker: Unit, defender: Unit) {
  it(
    "calculates " + attacker.name + " vs " + defender.name + " differential",
    () => {
      const actual = calculateDefenseDifferential(attacker, defender);
      expect(actual).toMatchSnapshot();
    }
  );
}

it("returns a damage number for combat", () => {
  const damageDealt = combat(humanSpearmen, humanSpearmen);
  expect(damageDealt).toBeGreaterThanOrEqual(0);
});

it("calculates human spearmen vs human spearmen differential", () => {
  const actual = calculateDefenseDifferential(humanSpearmen, humanSpearmen);
  expect(actual).toMatchSnapshot();
});

it("calculates elf spearmen vs human spearmen differential", () => {
  const actual = calculateDefenseDifferential(elfSpearmen, humanSpearmen);
  expect(actual).toMatchSnapshot();
});

it("calculates human spearmen vs elf spearmen differential", () => {
  const actual = calculateDefenseDifferential(humanSpearmen, elfSpearmen);
  expect(actual).toMatchSnapshot();
});

it("calculates orc spearmen vs human spearmen differential", () => {
  const actual = calculateDefenseDifferential(orcSpearmen, humanSpearmen);
  expect(actual).toMatchSnapshot();
});

it("calculates orc spearmen vs elf spearmen differential", () => {
  const actual = calculateDefenseDifferential(orcSpearmen, elfSpearmen);
  expect(actual).toMatchSnapshot();
});

it("calculates human spearmen vs orc spearmen differential", () => {
  const actual = calculateDefenseDifferential(humanSpearmen, orcSpearmen);
  expect(actual).toMatchSnapshot();
});

testCombatDifferential(elfSpearmen,orcSpearmen);
