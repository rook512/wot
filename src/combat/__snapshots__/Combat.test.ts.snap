// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculates Elven Spearmen vs Orcish Spearmen differential 1`] = `0.3`;

exports[`calculates elf spearmen vs human spearmen differential 1`] = `0.5`;

exports[`calculates human spearmen vs elf spearmen differential 1`] = `0.1`;

exports[`calculates human spearmen vs human spearmen differential 1`] = `0.3`;

exports[`calculates human spearmen vs orc spearmen differential 1`] = `0.1`;

exports[`calculates orc spearmen vs elf spearmen differential 1`] = `0.3`;

exports[`calculates orc spearmen vs human spearmen differential 1`] = `0.5`;
